Metaspace memory leak in ClassUtils

Let's make sure issue is not already fixed in latest builds first.

- [x] I verified I can reproduce this issue against [latest Integration Build of Eclipse SDK](https://download.eclipse.org/eclipse/downloads/)

Ever since the M1 went live, I am no longer able to run our test suite without the process getting shot down, due to a significant increase in required memory. After a two week investigation, it turned out that the problem was not caused by the _heapspace_, as initially thought, but rather the _metaspace_. 

The following graph illustrates the evolution of the metaspace while executing our test suite. I think the memory leak is quite obvious. Honorary mention to the "plateau", which is where the Swing test suite is executed, indicating that this leak is exclusive to SWT.

![MemLeak-01](https://github.com/eclipse-platform/eclipse.platform.ui/assets/70652978/b7d68b5e-3d98-4d8e-8b18-4318db3c1154)

The problem doesn't show up when testing _the same code_ against the 2024-03. While there are moments where up to 90000 classes are loaded, the process always drops down to 30000-40000 classes shortly after.

![MemLeak-01-2024-03](https://github.com/eclipse-platform/eclipse.platform.ui/assets/70652978/67842428-9bd6-4ccb-9fb1-84d947ef9fbf)

To briefly summarize how our tests (and WindowBuilder in general works): For each test, a subclass of an SWT widget is created. This can be a Shell, Table, Composite etc. After analyzing the heapdump corresponding to the first graph, I noticed that those classes are kept in memory via a reference to the CSS Engine, which is what led me here in the first place.

![MemLeak-HeapDump](https://github.com/eclipse-platform/eclipse.platform.ui/assets/70652978/ffb4fdde-1a52-4a3b-a03a-b2ff978e6c27)

To be more specific, the memory leak is introduced by https://github.com/eclipse-platform/eclipse.platform.ui/commit/4e9942b470908591e06ca6ee9c4bf10eb9efbc5f. Or to be even more concrete, the following line is a **very** bad idea:

> private static final Map<Class<?>, String> simpleNames = new ConcurrentHashMap<>();

As mentioned, WindowBuilder creates several subclasses for widgets. Because they are widgets, a CSS style is applied and via several steps, they end up in this map. Because each subclass is unique, a new entry is added to the map for each instance. Because this map is never cleared, those classes (and by extend all class loaded by the same classloader) remain in memory indefinitely, even though the tests that created them have long succeeded. 

I haven't tested yet whether the issue can be avoided by using a WeakHashMap but that's something I wanted to try within the next few days. If it doesn't, then I propose reverting the problematic revision, as it doesn't achieve the intended goal of improving memory utilization, but instead introduces a crippling memory leak.

To prove my point, this is what the memory looks like if I clear this cache after each test. I think the similarity to the 2024-03 graph is obvious:

![MemLeak-01-fix](https://github.com/eclipse-platform/eclipse.platform.ui/assets/70652978/adb7b4cc-9143-43a6-be63-c0092f46dbbf)

## Steps to reproduce

From a fresh installation and clean workspace:

- Checkout the [WindowBuilder](https://github.com/eclipse-windowbuilder/windowbuilder.git) project.
- Remove the following line in DesignerTestCase:
  - `((Map<?, ?>) ReflectionUtils.getFieldObject(ClassUtils.class, "simpleNames")).clear();`
- Execute the tests via the command line: `mvn clean verify -fae`.
- Attach a profiler to the Surefire process and observe the metaspace and/or the number of loaded classes.

## Here is some relevant log output

```
[ERROR] Failed to execute goal org.eclipse.tycho:tycho-surefire-plugin:4.0.4:test (default-test) on project org.eclipse.wb.tests: An unexpected error occurred while launching the test runtime (process returned error code 137(SIGKILL received?)). The process logfile /home/jenkins/agent/workspace/build_PR-750/org.eclipse.wb.tests/target/work/data/.metadata/.log might contain further details. Command-line used to launch the sub-process was /opt/tools/java/openjdk/jdk-17/latest/bin/java -Dosgi.noShutdown=false -Dosgi.os=linux -Dosgi.ws=gtk -Dosgi.arch=x86_64 -XX:MaxRAMPercentage=40.0 -Dosgi.clean=true -jar /home/jenkins/agent/workspace/build_PR-750/.m2/repository/p2/osgi/bundle/org.eclipse.equinox.launcher/1.6.800.v20240330-1250/org.eclipse.equinox.launcher-1.6.800.v20240330-1250.jar -data /home/jenkins/agent/workspace/build_PR-750/org.eclipse.wb.tests/target/work/data -install /home/jenkins/agent/workspace/build_PR-750/org.eclipse.wb.tests/target/work -configuration /home/jenkins/agent/workspace/build_PR-750/org.eclipse.wb.tests/target/work/configuration -application org.eclipse.tycho.surefire.osgibooter.uitest -testproperties /home/jenkins/agent/workspace/build_PR-750/org.eclipse.wb.tests/target/surefire.properties -testApplication org.eclipse.ui.ide.workbench -nl de -clearPersistedState -consoleLog in working directory /home/jenkins/agent/workspace/build_PR-750/org.eclipse.wb.tests -> [Help 1]
```

## Tested under this environment:
* OS & version: Linux
* Eclipse IDE/Platform version (as shown in *Help > About*): Eclipse 4.32 M1

## Community

- [x] I understand reporting an issue to this OSS project does not mandate anyone to fix it. Other contributors may consider the issue, or not, at their own convenience. The most efficient way to get it fixed is that I fix it myself and contribute it back as a good quality patch to the project.

--- a/bundles/org.eclipse.e4.ui.css.core/src/org/eclipse/e4/ui/css/core/utils/ClassUtils.java
+++ b/bundles/org.eclipse.e4.ui.css.core/src/org/eclipse/e4/ui/css/core/utils/ClassUtils.java
@@ -13,24 +13,17 @@
  *******************************************************************************/
 package org.eclipse.e4.ui.css.core.utils;
 
-import java.util.Map;
-import java.util.concurrent.ConcurrentHashMap;
-
 /**
  * Class utils.
  */
 public class ClassUtils {
 
-	private static final Map<Class<?>, String> simpleNames = new ConcurrentHashMap<>();
 	/**
-	 * Return a simple name of Class <code>c</code>. For inner classes, the hyphen
-	 * is used, e.g., for Outer$Inner, the return value is "Outer-Inner"
+	 * @param c the class for what a simple name is to be returned
+	 * @return a simple name of Class <code>c</code>. For inner classes, the hyphen
+	 *         is used, e.g., for Outer$Inner, the return value is "Outer-Inner"
 	 */
 	public static String getSimpleName(Class<?> c) {
-		return simpleNames.computeIfAbsent(c, ClassUtils::computeSimpleName);
-	}
-
-	private static String computeSimpleName(Class<?> c) {
 		String name = c.getName();
 		int index = name.lastIndexOf('.');
 		if (index > 0) {
@@ -42,9 +35,17 @@ public class ClassUtils {
 
 	/**
 	 * Return the package name of Class <code>c</code>.
+	 *
+	 * @param c the class for what a simple name is to be returned
+	 * @return the package name, if this class represents an array type, a primitive
+	 *         type or void, this method returns <code>null</code>.
 	 */
 	public static String getPackageName(Class<?> c) {
-		Package p = c.getPackage();
-		return p == null ? null : p.getName();
+		String name = c.getName();
+		int index = name.lastIndexOf('.');
+		if (index > 0) {
+			return name.substring(0, index).intern();
+		}
+		return null;
 	}
 }

